### ISSUE TITLE
ZeroAndOneInflatedBetaOutput: module 'mxnet.symbol' has no attribute 'logical_or'

### ISSUE THREAD

**Issue by deejlucas** [awslabs/gluon-ts#1211]
I originally thought this was a typo, but I tried to make the change to np.logical_or locally, and that did not fix the problem. Please let me know if more information would be helpful

In gluonts 0.6.3, I am submitting the following:
```
estimator_4 = deepar.DeepAREstimator(
    freq="W", prediction_length=4, trainer=trainer_1, distr_output=ZeroAndOneInflatedBetaOutput(),
    context_length=26, num_cells=104, dropout_rate=0.1)
predictor_4 = estimator_4.train(training_data=data_train_beta)
```
This produces the error referenced in the title with the following stack trace:
```
  0%|          | 0/50 [00:00<?, ?it/s]
learning rate from ``lr_scheduler`` has been overwritten by ``learning_rate`` in optimizer.
  0%|          | 0/50 [00:01<?, ?it/s]
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-24-94b4fecb46f1> in <module>
      2     freq="W", prediction_length=4, trainer=trainer_1, distr_output=ZeroAndOneInflatedBetaOutput(),
      3     context_length=26, num_cells=104, dropout_rate=0.1)
----> 4 predictor_4 = estimator_4.train(training_data=data_train_beta)

~/.conda/envs/cuda_test/lib/python3.6/site-packages/gluonts/model/estimator.py in train(self, training_data, validation_data, num_workers, num_prefetch, shuffle_buffer_length, **kwargs)
    274             num_prefetch,
    275             shuffle_buffer_length,
--> 276             **kwargs,
    277         ).predictor

~/.conda/envs/cuda_test/lib/python3.6/site-packages/gluonts/model/estimator.py in train_model(self, training_data, validation_data, num_workers, num_prefetch, shuffle_buffer_length, **kwargs)
    247             input_names=get_hybrid_forward_input_names(trained_net),
    248             train_iter=training_data_loader,
--> 249             validation_iter=validation_data_loader,
    250         )
    251 

~/.conda/envs/cuda_test/lib/python3.6/site-packages/gluonts/mx/trainer/_base.py in __call__(self, net, input_names, train_iter, validation_iter)
    343                     )
    344 
--> 345                     epoch_loss = loop(epoch_no, train_iter)
    346                     if is_validation_available:
    347                         epoch_loss = loop(

~/.conda/envs/cuda_test/lib/python3.6/site-packages/gluonts/mx/trainer/_base.py in loop(epoch_no, batch_iter, is_training)
    260                             if first_forward:
    261                                 first_forward = False
--> 262                                 _ = net(*inputs)
    263                                 if self.post_initialize_cb:
    264                                     self.post_initialize_cb(net)

~/.conda/envs/cuda_test/lib/python3.6/site-packages/mxnet/gluon/block.py in __call__(self, *args)
    680             hook(self, args)
    681 
--> 682         out = self.forward(*args)
    683 
    684         for hook in self._forward_hooks.values():

~/.conda/envs/cuda_test/lib/python3.6/site-packages/mxnet/gluon/block.py in forward(self, x, *args)
   1242                                      'Find all contexts = {}'.format(ctx_set))
   1243                 with ctx:
-> 1244                     return self._call_cached_op(x, *args)
   1245             with ctx:
   1246                 try:

~/.conda/envs/cuda_test/lib/python3.6/site-packages/mxnet/gluon/block.py in _call_cached_op(self, *args)
    993     def _call_cached_op(self, *args):
    994         if self._cached_op is None:
--> 995             self._build_cache(*args)
    996         assert self._cached_op, "Gluon failed to build the cache. " \
    997                                 "This should never happen. " \

~/.conda/envs/cuda_test/lib/python3.6/site-packages/mxnet/gluon/block.py in _build_cache(self, *args)
    922 
    923     def _build_cache(self, *args):
--> 924         data, out = self._get_graph(*args)
    925         data_names = {data.name: i for i, data in enumerate(data)}
    926         params = self.collect_params()

~/.conda/envs/cuda_test/lib/python3.6/site-packages/mxnet/gluon/block.py in _get_graph(self, *args)
    914             params = {i: j.var() for i, j in self._reg_params.items()}
    915             with self.name_scope():
--> 916                 out = self.hybrid_forward(symbol, *grouped_inputs, **params)  # pylint: disable=no-value-for-parameter
    917             out, self._out_format = _flatten(out, "output")
    918 

~/.conda/envs/cuda_test/lib/python3.6/site-packages/gluonts/model/deepar/_network.py in hybrid_forward(self, F, feat_static_cat, feat_static_real, past_time_feat, past_target, past_observed_values, future_time_feat, future_target, future_observed_values)
    456 
    457         # (batch_size, seq_len)
--> 458         loss = distr.loss(target)
    459 
    460         # (batch_size, seq_len, *target_shape)

~/.conda/envs/cuda_test/lib/python3.6/site-packages/gluonts/mx/distribution/distribution.py in loss(self, x)
    136             for each event in `x`.
    137         """
--> 138         return -self.log_prob(x)
    139 
    140     def prob(self, x: Tensor) -> Tensor:

~/.conda/envs/cuda_test/lib/python3.6/site-packages/gluonts/mx/distribution/transformed_distribution.py in log_prob(self, y)
    147             y = x
    148 
--> 149         return self.base_distribution.log_prob(x) + lp
    150 
    151     def cdf(self, y: Tensor) -> Tensor:

~/.conda/envs/cuda_test/lib/python3.6/site-packages/gluonts/mx/distribution/inflated_beta.py in log_prob(self, x)
     80 
     81         # mask zeros for the Beta distribution input to prevent NaN gradients
---> 82         inputs = F.where(F.logical_or(x == 0, x == 1), x.zeros_like() + 0.5, x)
     83 
     84         # compute log density, case by case

AttributeError: module 'mxnet.symbol' has no attribute 'logical_or'
```

**Comment by lostella** [comment]
Looks like `mx.nd` has it while `mx.sym` doesn't have it. Maybe a fix is to use `broadcast_logical_or`, which appears to be in both

**Comment by lostella** [comment]
@deejlucas do you want to take a shot at this and open a PR?

**Comment by deejlucas** [comment]
@lostella Sure, I'll take a shot at it.

**Comment by deejlucas** [comment]
@lostella I've made this simple change and verified that my code runs. A couple questions about testing for you or anyone else who knows.

1. There are a set of tests that fail even when running pytest in master (including all tests in `test/distribution/test_flows.py`). Is this expected? Is it okay not to concern myself with tests that don't cover this change?
2. Now that the code is running, I get NA losses for all batches in the first epoch when I try to train a model on the data that originally motivated me to try this out. This happens for all sets of hyperparameters in a random search of a reasonable hyperparameter space. This is also the case when I try to use a set of state unemployment figures as the targets (something that seemed potentially appropriate based on literature). This may be outside of the scope of the issue, but I'm hesitant to put in a PR without training a model that actually works. Do you know of a dataset that is known to be suitable for this distribution?

**Comment by lostella** [comment]
@deejlucas 

1. Yes that's fine, there are some tests which occasionally fail due to randomness, some of these could use some improvement but you don't have to worry about it.

2. You could try with the `"traffic"` dataset, which should have data in [0, 1].

**Comment by deejlucas** [comment]
@lostella 
It seems like something is wrong with the log_prob function in the underlying beta distribution, and I'm afraid my ability to troubleshoot symbolic programming is coming up short. 

The code below produces nan loss in the first epoch. I tried removing elements of the `log_prob` functions in `Beta` and determined that [removing `F.log(1 - x)` prevents NaNs](https://github.com/awslabs/gluon-ts/blob/1b6109f34b6aa33c11bd9d83be7fd389ec74d7ec/src/gluonts/mx/distribution/beta.py#L67). This is confusing to me, as there's no reason I can think of that `F.log(1-x)` should be problematic when `F.log(x)` is not.

Should we consider this outside the scope of this issue? I've addressed the error that originally caused me to open this issue. I can do a PR of the trivial fix for that and open a new one for the issue described above, but it would be a PR to fix broken code, unless I'm really mixed up here.

If this is out of scope for this issue, I'd be happy to work on the new issue too, if someone a little more knowledgeable can point me in the right direction.

```
import numpy as np
import pandas as pd

from gluonts.dataset import common
from gluonts.model import deepar
from gluonts.mx.trainer import Trainer
from gluonts.mx.distribution.beta import BetaOutput

start = pd.to_datetime('1999-01-04')

train = []
for d in range(10000):
    train.append({"target": np.random.beta(0.1, 0.25, 1000), "start": start})

dataset = common.ListDataset(train, freq="W-MON", one_dim_target=True)

#%%
import mxnet as mx

trainer_1 = Trainer(ctx=mx.context.gpu(),
                      epochs=100,
                      batch_size=32,
                      learning_rate=1e-8,
                      minimum_learning_rate=5e-10,
                      weight_decay=1e-8
                    )

estimator_3 = deepar.DeepAREstimator(
    freq="W", prediction_length=24, trainer=trainer_1, distr_output=BetaOutput(),
    context_length=48, num_cells=96, dropout_rate=0.1
    )

predictor_3 = estimator_3.train(training_data=dataset)
```

**Comment by lostella** [comment]
@deejlucas does the problem arise on CPU too?

**Comment by deejlucas** [comment]
@lostella I get the same result (much more slowly :))

**Comment by lostella** [comment]
@deejlucas you can set `hybridize=False` and debugging will be much easier. I'm looking at it and indeed something looks off with `Beta`

**Comment by lostella** [comment]
@deejlucas found the issue, nothing wrong with `Beta` :-) you need to turn off the target scaling in the estimator:

```python
DeepAREstimator(..., scaling=False)
```

**Comment by deejlucas** [comment]
Ah, that makes perfect sense! I'll get in the PR for this issue then. Thanks for taking a look at this with me

**Comment by lostella** [comment]
One note: I had to clip the target value because otherwise observations very close to 0 or 1 would still raise nans in your example. So I did it like this:

```python
train = []
for d in range(100):
    target = np.random.beta(0.1, 0.25, 1000)
    target = np.clip(target, 1e-7, 1. - 1e-7)
    train.append({"target": target, "start": start})
```

This might be something that the distribution could handle internally when computing `log_prob`. On the other hand, one probably wants anything >=1 or <=0 to yield nans (e.g. to spot the scaling issue above), so blindly clipping within `log_prob` may not be ideal.
